Lib/email/headerregistry.py

```diff
@@ -99,6 +105,10 @@ def __eq__(self, other):
                 self.username == other.username and
                 self.domain == other.domain)
 
+    def _validate_part(self, value):
+        """Parts cannot contain CRLF for security reasons."""
+        if '\r\n' in value:
+            raise ValueError("invalid address; address parts cannot contain CRLF")
```

Comment:
The email package's style is to put functions before where they are used (but see below).

The input should be checked at the top of the method.  get_addr_spec expects all CRLF to have already been removed from its input, so it won't be checking for it either, and I don't know off the top of my head if it will fail.  Regardless of that, though, the safe thing to do is check all input strings at the start of the function.

In addition, you should check for CR and LF independently, since either one could get interpreted as if it were CRLF at the receiving end.  I would recommend something like (untested):

    inputs = ''.join(filter(None, (display_name, username, domain, addr_spec))
    if '\r' in inputs or '\n' in inputs:
        raise [...]

It might be nicer to do it in a loop and indicate which argument is in error, but either way I'd do it without a separate function.

Reply:
Done.

Reply:
I'm curious, do you know of any http libraries that interpret CR / LF independently as if it were CRLF?

Reply:
I don't, but the python email package at least does so for CR, and I'm pretty sure there are other email packages that do so as well, since it is the standard linux line-end.  Checking LF independently as well is just a belt-and-suspenders move :)